pkg/controller/cronjob/cronjob_controllerv2.go

```diff
@@ -0,0 +1,647 @@
+		// TODO: @alpatel: confirm with @soltysh the following TODO is not true anymore. We will now
+		//      only requeue the for the next scheduled time, instead of hitting this error again
+		//      and again. With the new workflow we might only hit this if controller got wedged
+		//		and on restart we miss the schedule time by more than deadline. In that case
+		// 		schedule for next time
+		// TODO: Since we don't set LastScheduleTime when not scheduling, we are going to keep noticing
+		// the miss every cycle.  In order to avoid sending multiple events, and to avoid processing
+		// the cj again and again, we could set a Status.LastMissedTime when we notice a miss.
+		// Then, when we call getRecentUnmetScheduleTimes, we can take max(creationTimestamp,
+		// Status.LastScheduleTime, Status.LastMissedTime), and then so we won't generate
+		// and event the next time we process it, and also so the user looking at the status
+		// can see easily that there was a missed execution.
+		t := nextScheduledTimeDuration(sched, now)
+		return t, nil
+	}
+	if cj.Spec.ConcurrencyPolicy == batchv1beta1.ForbidConcurrent && len(cj.Status.Active) > 0 {
+		// Regardless which source of information we use for the set of active jobs,
+		// there is some risk that we won't see an active job when there is one.
+		// (because we haven't seen the status update to the SJ or the created pod).
+		// So it is theoretically possible to have concurrency with Forbid.
+		// As long the as the invocations are "far enough apart in time", this usually won't happen.
+		//
+		// TODO: @alpatel confirm we @soltysh we can probably set a deterministic job name per
+		// 		schedule time. The formula: last_run=UTCInSeconds(now)-UTCInSeconds(creationTime)/intervalInSeconds
+		//  	might give use last scheduled counter. May be save it in status and reconcile on that
+		// TODO: for Forbid, we could use the same name for every execution, as a lock.
+		// With replace, we could use a name that is deterministic per execution time.
+		// But that would mean that you could not inspect prior successes or failures of Forbid jobs.
+		klog.V(4).InfoS("Not starting job because prior execution is still running and concurrency policy is Forbid", "cronjob", klog.KRef(cj.GetNamespace(), cj.GetName()))
+		t := nextScheduledTimeDuration(sched, now)
+		return t, nil
+	}
+	if cj.Spec.ConcurrencyPolicy == batchv1beta1.ReplaceConcurrent {
+		for _, j := range cj.Status.Active {
+			klog.V(4).InfoS("Deleting job that was still running at next scheduled start time", "job", klog.KRef(j.Namespace, j.Name))
+
+			job, err := jc.GetJob(j.Namespace, j.Name)
+			if err != nil {
+				recorder.Eventf(cj, corev1.EventTypeWarning, "FailedGet", "Get job: %v", err)
+				return nil, err
+			}
+			if !deleteJob(cj, job, jc, recorder) {
+				return nil, fmt.Errorf("could not replace job %s/%s", job.Namespace, job.Name)
+			}
+		}
+	}
+
+	jobReq, err := getJobFromTemplate(cj, scheduledTime)
```

Comment:
make clear in the godoc on this function and here that having deterministic names is critical to proper functioning of the controller to avoid accidentally creating multiple jobs for the same cronjob,schedule-time tuple